File: algo/lc.202.cc
class Solution {
public:
    bool isHappy(int n) {
        map<int, bool> visited;
        visited[n] = true;
        int prev = n, next = 0;
        while (prev != 1) {
            while (prev > 0) {
                next += (prev%10)*(prev%10);
                prev /= 10;
            }
            // next == 1 ?
            if (next == 1) return true;
            // visited next ?
            if (visited.find(next) != visited.end()) {
                return false; // cycle detected
            }
            // add to map and clean up
            visited[next] = true;
            prev = next;
            next = 0;
        }
        return true;
    }
};

File: algo/lc.202.py
class Solution:
    def isHappy(self, n: int) -> bool:
        previous = []
        x = n
        while True:
            sumsq = sum(int(x)**2 for x in str(x))
            if sumsq == 1:
                return True
            else:
                if sumsq in previous:
                    return False
                else:
                    previous.append(sumsq)
                    x = sumsq
